File: Kattis/Java/EncodedMessage.java
package encodedmessage;

import java.util.Scanner;

public class EncodedMessage {

    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        int cases = input.nextInt();
        
        for (int i = 0; i <= cases; i++) {
            String encodedMessage = input.nextLine();
            int length = (int) Math.sqrt(encodedMessage.length());
            DecodeMessage(length, encodedMessage);
        }
    }

    private static void DecodeMessage(int length, String encodedMessage) {
        for (int j = 0; j < length; j++) {
            for (int k = (length - 1); k < (encodedMessage.length()); k += length) {
                System.out.print(encodedMessage.charAt(k - j));
            }
        }
        System.out.println();
    }
    
}

File: Kattis/CPlusPlus/EncodedMessage.cpp
#include <iostream>
#include <cmath> 

void DecodeMessage (int, std::string);

int main() {
    int cases, length;
    std::string encodedMessage;

    std::cout << "How many messages would you like to decrypt? ";
    std::cin >> cases;

    for (int i = 0; i < cases; i++) {
        std::cout << "Enter the message: ";
        std::cin >> encodedMessage;
        length = sqrt(encodedMessage.length());
        DecodeMessage(length, encodedMessage);
    }

    system("PAUSE");
    return 0;
}

void DecodeMessage (int length, std::string encodedMessage) {
    std::cout << "Decrypted Message: ";
    for (int j = 0; j < length; j++) {
        for (int k = (length - 1); k < (encodedMessage.length()); k+= length) {
            std::cout << encodedMessage.at(k - j);
        }
    }
    std::cout << std::endl;
}